### ISSUE TITLE
CommandList Parameters (thoughts)

### ISSUE THREAD

**Issue by SinsOfSeven** [bo3b/3Dmigoto#342]
Recently, I've been making a LOT of "libraries" for 3dmigoto (GIMI).

Src of said "Libs"
[3dmigoto_math_lb](https://github.com/SinsOfSeven/3dmigoto_math_lib) - an incomplete proof of concept for doing more typical math functions in 3dmigoto.
[SliderImpact](https://github.com/SinsOfSeven/SliderImpact) - A set of loosely related tools based around the concept of custom slider UIs, like making shapekeys by editing the position information of a buffer in a computer shader, or generating a virtual texture based on the slider positions which can be used as a namespaced resource.
[TexFx](https://github.com/SinsOfSeven/TexFx) - A specialized tool for creating transparency and soon other special effects such as parallax or animated effects.

By abusing namespaced variables, I have effectively created functions with parameters, but while I was exploring the 3dmigoto source code, I noticed that in the section dedicated to declaring variables, there was a comments about reserving `$[0-9]` variables for commandlist parameters. I tried to think of how to implement it, but I've been avoiding anything that involves editing the src. Mostly because I'm out of my depth. 

I was kind of curious of how this would be implemented though. 
```ini
namespace = ex

[CommandListF]
local $1param
local $2param
local $constX = 5
return = ($param1 + $param2) * $constX

[CommandListSomewhereElse]
x = run = CommandList\ex\F, 1,3
; x = 20
```

But this introduces the idea of having a return type for commandlists, so maybe we can ignore that.

```ini
namespace = ex

[Constants]
$return

[CommandListF]
local $1param
local $2param
$return = ($param1 + $param2)

[CommandListSomewhereElse]
run = CommandList\ex\F, 3, 6
x = $\ex\return
```

I think this would be acceptable, since I think most "Functions" wouldn't have a return in 3dmigoto.

```ini
[CommandListCheckClick]
local $1_xoffset
local $2_xsize
local $3_yoffset
local $4_ysize
if cursor_x > $1_xoffset && cursor_x < $1_xoffset + $2x_size
   if cursor_y > $3_yoffset && cursor_y < $3_yoffset + $4y_size
    ;cursor is inside the screen region, do thing
  endif
endif

[Present]
run = CommandListCheckClick, 0.4 ,0.2, 0.1, 0.4
run = CommandListCheckClick, 0.6 ,0.1, 0.9, 0.5
```

Also, I realized that one of the most useful places for this is just in a `[CustomShader]`, so it would be nice to have them there.

```ini
[CustomShaderF]
local $1_DrawNum
local $2_DrawOffset
;shaderstuff
drawindexed = $1_DrawNum, $2_DrawOffset, 0

[Present]
if $condition
  ; set buffers and textures
  run = CustomShader, 24,120
endif
```

These next two are how I've used it in a real mod.

```ini
; Mod Example INI
namespace = texfx

[Constants]
global $_1 = 0
global $_2 = 0

[CustomShaderTransparencySpecial1]
handling = skip
cull = none
run = CommandListHullHack
vs = OutlineHull.hlsl
ps = OutlineWithDiffuseColor1.hlsl
;drawindexed = auto
drawindexed = $_1, $_2, 0
$_1 = 0
$_2 = 0
ps-t69 = null
```
```ini
[TextureOverrideBody]
hash = 1234abcd
match_first_index = 13077
match_priority = 0
ib = ResourceBodyIB
ps-t0 = ResourceBodyNormalMap
ps-t1 = ResourceBodyDiffuse
ps-t2 = ResourceBodyLightMapEdit
; Texture order fix for some shaders
run = commandlist\or\fix
;draw main body
drawindexed = 36420, 0, 0
if $HairOrniment == 1
    drawindexed = 1224, 36420, 0
endif
if $EarRings == 1
    drawindexed = 990, 37644, 0
endif
if $Hood == 1
    ps-t1 = ResourceBodyDiffuseEdit
    ;run = commandlist\or\fix
    ps-t69 = Resource\texfx\transparencyMask
    if ps == 037730.0 && $\texfx\texfx_on
        $\texfx\_1 = 2394
        $\texfx\_2 = 38634
        run = CommandList\TexFx\SetResourceReferences
        run = CustomShader\texfx\TransparencySpecial0
    endif
    ;drawindexed = 2394, 38634, 0
    run = CommandList\TexFx\ResetResources
endif
```

Another example that's a little

```ini
[CommandListComputeShapeKeys]
cs-u5 = copy ResourceColleiPositionBase
run = CommandListKey1
run = CommandListKey2
run = CommandListKey3
ResourcePosition = copy cs-u5
; clean up
x88 = 0
cs-u5 = null
cs-t50 = null
cs-t51 = null

[CommandListKey1]
cs-t50 = copy ResourceColleiPositionBase
cs-t51 = copy ResourceColleiPositionBooba
x88 = $Key1
run = CustomShaderKeys

[CustomShaderKeys]
cs = resources/ShapeKey.hlsl
dispatch =  16910, 1, 1

;Note the $Key1 is set by a UI slider somewhere else.
```

Rambling ->

I also noticed that a lot of the things we're doing would be very appropriate for looping, but I'm hesitant to bring up looping. I actually made a recursion "loop" when I noticed that I got an error about self referencing, and I don't really want to think  about it being part of the feature set.

Even if  we can't explicitly loop through them, we've also run into many issues where having an array or something similar would help.
I was wondering how hard it would be to make it possible to read values by index from a custom Resource. Writing to a custom resource isn't too hard, it can be done in a compute shader. But reading them back has proven to be really annoying, or costly. SilentNightSound made a modification for reading back a value from a constant buffer from the GPU, but I think it might  be easier to just copy the buffer `ResourceCBCopy = copy cb1` then read a value, but I don't know if this is slow or not, and mostly I would want to do it for custom information anyways.

I would like to clarify this post isn't being made with any expectations. I'm very happy overall with what we've been able to accomplish, and I don't want to introduce new annoying things. I'm also curious if there are any features I've already overlooked. I'm definitely not an expert on 3dmigoto. I'm sorry for the things I've done to this very splendid 3d program.

This reminds me of when I tried `run = KeyX` and got sad it didn't just work, I just wanted an easy way to generate sine waves before I made that math_lib.

**Comment by DarkStarSword** [comment]
Sorry for the delay in replying, this has been in the back of my mind to give this some thought, I've just been a bit busy.

> By abusing namespaced variables, I have effectively created functions with parameters, but while I was exploring the 3dmigoto source code, I noticed that in the section dedicated to declaring variables, there was a comments about reserving `$[0-9]` variables for commandlist parameters. I tried to think of how to implement it, but I've been avoiding anything that involves editing the src. Mostly because I'm out of my depth.
> 
> I was kind of curious of how this would be implemented though.

I believe when I wrote that I would have been thinking along the lines of how shell scripts or perl passes parameters to functions - inside the function $1 is the first positional parameter, $2 is the 2nd, and so on. The exact syntax is TBD (e.g. do we use brackets to pass parameters like most languages, or just use commas or spaces as delimiters like shell or AHKv1?), but it could look something like this:

```
[CommandListF]
local $constX = 5
$return = ($1 + $2) * $constX

[CommandListSomewhereElse]
run = CommandListF(1, 3)
```

There's usually a few extra special variables, like in shell "$*" is all arguments together (or "$@" which also includes the script name in $0), $# is the number of arguments, etc.

The downside of languages like shell or perl is that the arguments aren't named when passed like this, so there can be some confusion as to what argument is what, and adding/changing arguments is harder to do without breaking backwards compatibility, though usually these only become a concern as the program becomes more complex. I can't really see a good clean way to add fully named arguments to our existing syntax, but sometimes shell scripts do something like this to name the arguments internally, which might be a reasonable compromise:
```
[CommandListF]
local $first_named_arg = $1
local $second_named_arg = $2
...
```

> But this introduces the idea of having a return type for commandlists, so maybe we can ignore that.

I haven't given much thought to return values. The easiest implementation is probably just to have a global return variable like your example, which would probably suffice for most cases, and if someone needs to preserve a return value they can just copy it into a local/global variable. I will say though, I do like languages that allow a function to return multiple values, either in a tuple form like Python (e.g. `return (a, b, c)`), or by having some parameters be output variables. Adding tuples is probably too much work, but output variables are probably doable. Might look something like this:

```
[CommandListF]
local $constX = 5
$3 = ($1 + $2) * $constX

[CommandListSomewhereElse]
local $ret
run = CommandListF(1, 3, $ret)
```

Downside of this is that there is no formal definition or hint that $3 is an output variable here, and that's generally frowned upon in most language design (a function unexpectedly changing a variable is a recipe for subtle bugs). Sometimes an "out" keyword is used for this, but given the existing syntax the only clean place we could add it is at the caller... I guess maybe we could also add it at the assignment and then throw an error if this pattern is used and one or both of the out keywords is missing:

```
[CommandListF]
local $constX = 5
out $3 = ($1 + $2) * $constX

[CommandListSomewhereElse]
local $ret
run = CommandListF(1, 3, out $ret)
```

> Also, I realized that one of the most useful places for this is just in a `[CustomShader]`, so it would be nice to have them there.

Agreed.

> Rambling ->
> 
> I also noticed that a lot of the things we're doing would be very appropriate for looping, but I'm hesitant to bring up looping. I actually made a recursion "loop" when I noticed that I got an error about self referencing, and I don't really want to think about it being part of the feature set.

So I kind of intentionally didn't add looping yet, because it's easy to add a lot of work in a loop that runs on every draw call and kill performance, or accidentally add an infinite loop which would manifest as a game freeze - so, in some ways I guess I was hoping that the limitation would lead to creative solutions that didn't have the same risks. In a lot of cases recursion can be used as a substitute for a loop, and 3DMigoto intentionally has a hard recursion limit (`#define MAX_COMMAND_LIST_RECURSION 64`) so although you could still end up harming performance with that, it should never be able to hang entirely (and also that limit is there to make sure you never end up overflowing the actual stack of the game as that could lead to a crash).

But that's not to say that we would never add looping, even if it's just a for each style loop as that's harder to accidentally make an infinite loop with those. If we add positional arguments to functions it would make a lot of sense to be able to loop over them.

> Even if we can't explicitly loop through them, we've also run into many issues where having an array or something similar would help.

@davegl1234 added arrays when he did the DX9 port (e.g. see [here,](https://github.com/bo3b/3Dmigoto/blob/master/DirectX9/CommandList.cpp#L4903), and I've been meaning to take a closer look at exactly what he did and see about porting them back to DX11. (@davegl1234 I'm at work at the moment and can't check the geo-11 source code - did you by any chance port them back to DX11 in that fork)?

> I was wondering how hard it would be to make it possible to read values by index from a custom Resource. Writing to a custom resource isn't too hard, it can be done in a compute shader. But reading them back has proven to be really annoying, or costly. SilentNightSound made a modification for reading back a value from a constant buffer from the GPU, but I think it might be easier to just copy the buffer `ResourceCBCopy = copy cb1` then read a value, but I don't know if this is slow or not, and mostly I would want to do it for custom information anyways.

The problem here is that the GPU can be up to several frames behind the CPU, so you can't sit and wait for the value or you will stall the game well beyond the point of unplayability. As much as possible this scenario should be avoided by design and the data flow should be one way from CPU to GPU only, though in situations where having a value from some unpredictable number of frames ago is acceptable it can be done - currently there is code to do this to support the "automatic convergence" feature (search CommandList.cpp for "staging_type" - currently it only applies to convergence/separation assignments), and I always meant to expand that to allow retrieving values from other (and perhaps arbitrary) buffers as well. Note that Dave made some modifications to the auto-convergence in geo-11 (my implementation in 3DMigoto used a CustomShader, his version builds the functionality into geo-11 itself), and I'm not sure if he made any changes to the staging code we should bring in for this.

> This reminds me of when I tried `run = KeyX` and got sad it didn't just work, I just wanted an easy way to generate sine waves before I made that math_lib.

I'm open to requests if there are any useful functions you would like added built in. sin/cos is certainly doable.

**Comment by SinsOfSeven** [comment]
No worries about the time to reply. I don't think that any of this is a requirement, sometimes it's more fun to work inside constraints. Something we've been doing lately is drawing UI elements, and I've been trying to think up ways to reduce the number of calls required to draw them. I'm not a huge fan of every single element needing it's own call, but I also can't think of a way (in the bounds of 3dmigoto) to make it less troublesome.

I had taken some time when thinking about adding the math functions, and I started to wonder what built in functions like that could look like.

For example if we wanted Cos(), I kind of assume it would work like the current operator for floor() and fmod() 
`//` and `%`
So what would be a way to do any named function like that, `_COS 5` ? It's kind of fun to think about. I've had a blast working with 3dmigoto. I know that the amount of time spent on it isn't trivial, but it still surprises me just how much we can do with it. It really does feel like simple scripting for directx rendering. I think that one of the next things I want to take on is trying to learn how to use DrawInstanced to make particle generators or something. we're having a blast using segmented draw instructions to make parts with unique textures and shaders. We've been calling them components. That was really fun to think about too, we export models piecewise in blender then combine them and count up the verts and indices along the way to get the draw segements for each part, which can then be used for toggles, or changing the drawcall per part.

This is one of the examples where arrays came to mind, I made an encoder to store data in the d3dx_user.ini (persist). I cut down the example some, because there were 4 more of each of these. We noticed that 3dmigoto will sometimes struggle to load many persist vars, it might not have had too much trouble with 16, but I spent time making this to potentially fix an issue with another mod.
```ini
[Constants]
;Encoded Persist
global persist $color0
;Runtime Vars
global $r = 0
global $g = 0.5
global $b = 0
global $a = 0
post run = CommandListUnpack

;Runs once on load
[CommandListUnpack]
local $encode = 64
local $x = $encode - 1
local $temp = 0
$temp = $color0
$r = $temp%$encode/$x
$temp = $temp//$encode
$g = $temp%$encode/$x
$temp = $temp//$encode
$b = $temp%$encode/$x
$temp = $temp//$encode
$a = $temp%$encode/$x

; Run this whenever the values are updated.
[CommandListPack]
local $encode = 64
$color0 = ($r*0.9999*$encode//1) + ($g*0.9999*$encode//1*$encode) + ($b*0.9999*$encode//1*$encode**2) + ($a*0.9999*$encode//1*$encode**3)
```

The other mod uses a staggering 3800 persist vals. most of which are just binary flags. I wanted to know if there is anything I should keep in mind that might be technically challenging for 3dmigoto to do the functions like this, where I unpack the values on load into individual values, or if it might be better to do in-place operations on the encoded version of the val. 
I can only safely encode 23 or 24 bin flags into a float, so even doing it that way still requires a significant number of persist vals. (~160 persist which is better but still a lot.) And writing the functions to access those is tedious. Though computationally it's not very difficult. Maybe there is a smarter way to design this mod so it doesn't have such unrealistic usage, But It's still a little much haha.

I don't think it's that important to come up with solutions for this in 3dmigoto. I'm sharing them mostly out of professional curiosity, wondering what you think of these things. I've found it more than a little difficult to find other examples of people really pushing 3dmigoto. I'm trying to learn more about it so I can use it in new ways, So if you know of standout projects related to it, feel free to point them out. I recently joined the 3dvision community under this same username. That might be a more appropriate place to info dump on me.

As always
